File: agents/weekly-team-update/scripts/fetch-data.py
#!/usr/bin/env python3
"""Fetch weekly team data from GitHub, GitLab, and Jira APIs."""

import json
import os
import re
import sys
import urllib.request
import urllib.parse
import urllib.error
import ssl
from base64 import b64encode
from datetime import datetime, timedelta
from pathlib import Path

SCRIPT_DIR = Path(__file__).parent
DATA_DIR = SCRIPT_DIR.parent / "data"
CACHE_DIR = DATA_DIR / "cache"
CONFIG_PATH = DATA_DIR / "team-config.json"

TODAY = datetime.now().strftime("%Y-%m-%d")
SEVEN_DAYS_AGO = (datetime.now() - timedelta(days=7)).strftime("%Y-%m-%d")
SEVEN_DAYS_AGO_ISO = f"{SEVEN_DAYS_AGO}T00:00:00Z"

# Load config
with open(CONFIG_PATH) as f:
    config = json.load(f)

engineers = config["engineers"]
jira_config = config["jira"]
products = config["products"]
sprint_pattern = re.compile(config["sprint_name_pattern"])

# Tokens
GITHUB_PAT = os.environ.get("GITHUB_PAT", "")
GITLAB_PAT = os.environ.get("GITLAB_PAT", "")
JIRA_API_TOKEN = os.environ.get("JIRA_API_TOKEN", "")
JIRA_EMAIL = os.environ.get("JIRA_EMAIL", "")

ssl_ctx = ssl.create_default_context()


def github_request(url):
    req = urllib.request.Request(url)
    req.add_header("Authorization", f"token {GITHUB_PAT}")
    req.add_header("Accept", "application/vnd.github.v3+json")
    try:
        with urllib.request.urlopen(req, context=ssl_ctx) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        print(f"  GitHub error {e.code}: {url[:80]}", file=sys.stderr)
        return {"items": []}


def gitlab_request(url):
    req = urllib.request.Request(url)
    req.add_header("PRIVATE-TOKEN", GITLAB_PAT)
    try:
        with urllib.request.urlopen(req, context=ssl_ctx, timeout=15) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()[:100]
        print(f"  GitLab error {e.code}: {body}", file=sys.stderr)
        return []
    except (urllib.error.URLError, TimeoutError) as e:
        print(f"  GitLab connection error: {e}", file=sys.stderr)
        return []


def jira_request(jql, fields):
    base = f"https://{jira_config['cloud_id']}/rest/api/3/search/jql"
    params = urllib.parse.urlencode({
        "jql": jql,
        "maxResults": 100,
        "fields": ",".join(fields),
    })
    url = f"{base}?{params}"
    credentials = b64encode(f"{JIRA_EMAIL}:{JIRA_API_TOKEN}".encode()).decode()
    req = urllib.request.Request(url)
    req.add_header("Authorization", f"Basic {credentials}")
    req.add_header("Accept", "application/json")
    try:
        with urllib.request.urlopen(req, context=ssl_ctx) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        body = e.read().decode()[:200]
        print(f"  Jira error {e.code}: {body}", file=sys.stderr)
        return {"issues": []}


def csv_escape(value):
    value = str(value).replace('"', '""')
    if "," in value or '"' in value:
        return f'"{value}"'
    return value


def extract_issue_refs(body):
    if not body:
        return ""
    refs = set()
    for m in re.findall(
        r"(?:(?:closes?|fixes?|resolves?)\s*#(\d+))|(?:#(\d+))|(?:/issues/(\d+))",
        body,
        re.IGNORECASE,
    ):
        for g in m:
            if g:
                refs.add(g)
    return " ".join(sorted(refs))


def extract_sprint_name(sprint_field):
    if not sprint_field or not isinstance(sprint_field, list):
        return ""
    # Try active sprint matching pattern first
    for s in sprint_field:
        if isinstance(s, dict) and s.get("state") == "active":
            name = s.get("name", "")
            if sprint_pattern.match(name):
                return name
    # Any active sprint
    for s in sprint_field:
        if isinstance(s, dict) and s.get("state") == "active":
            return s.get("name", "")
    # Future sprint
    for s in sprint_field:
        if isinstance(s, dict) and s.get("state") == "future":
            return s.get("name", "")
    return ""


# --- GITHUB ---
print("=== Fetching GitHub PRs ===")
github_prs = []
seen_pr_keys = set()

for eng in engineers:
    name = eng["name"]
    username = eng["github"]
    print(f"  {name} ({username})...")

    # Merged PRs
    query = f"author:{username} is:pr is:merged merged:{SEVEN_DAYS_AGO}..{TODAY}"
    encoded = urllib.parse.quote(query)
    data = github_request(
        f"https://api.github.com/search/issues?q={encoded}&per_page=100"
    )
    for pr in data.get("items", []):
        key = (pr.get("repository_url", ""), pr["number"])
        repo_url = pr.get("repository_url", "")
        repo = "/".join(repo_url.split("/")[-2:]) if repo_url else ""
        merged_at = pr.get("pull_request", {}).get("merged_at", "") if pr.get("pull_request") else ""
        github_prs.append({
            "engineer": name,
            "number": pr["number"],
            "title": pr["title"],
            "repo": repo,
            "state": "merged",
            "created_at": pr.get("created_at", ""),
            "merged_at": merged_at,
            "html_url": pr.get("html_url", ""),
            "issue_refs": extract_issue_refs(pr.get("body", "")),
        })
        seen_pr_keys.add(key)

    # Open PRs
    query = f"author:{username} is:open is:pr"
    encoded = urllib.parse.quote(query)
    data = github_request(
        f"https://api.github.com/search/issues?q={encoded}&per_page=100"
    )
    for pr in data.get("items", []):
        key = (pr.get("repository_url", ""), pr["number"])
        if key in seen_pr_keys:
            continue
        repo_url = pr.get("repository_url", "")
        repo = "/".join(repo_url.split("/")[-2:]) if repo_url else ""
        github_prs.append({
            "engineer": name,
            "number": pr["number"],
            "title": pr["title"],
            "repo": repo,
            "state": "open",
            "created_at": pr.get("created_at", ""),
            "merged_at": "",
            "html_url": pr.get("html_url", ""),
            "issue_refs": extract_issue_refs(pr.get("body", "")),
        })
        seen_pr_keys.add(key)

print(f"  Total GitHub PRs: {len(github_prs)}")

# Save GitHub CSV
with open(CACHE_DIR / "github-prs.csv", "w") as f:
    f.write("engineer,number,title,repo,state,created_at,merged_at,html_url,issue_refs\n")
    for pr in github_prs:
        f.write(
            f"{csv_escape(pr['engineer'])},{pr['number']},{csv_escape(pr['title'])},"
            f"{pr['repo']},{pr['state']},{pr['created_at']},{pr['merged_at']},"
            f"{pr['html_url']},{pr['issue_refs']}\n"
        )

# --- JIRA ---
print("\n=== Fetching Jira Tickets ===")
jira_tickets = {}
active_sprint_name = ""

projects_clause = ", ".join(f'"{p}"' for p in jira_config["projects"])
fields = [
    "summary", "status", "assignee", "resolution", "resolutiondate",
    "statuscategorychangedate", "issuetype", "priority", "updated",
    "created", "customfield_10470", "customfield_10020", "issuelinks",
]

for eng in engineers:
    name = eng["name"]
    jira_id = eng["jira_account_id"]
    print(f"  {name} ({jira_id[:20]}...)...")

    jql = (
        f'(assignee = "{jira_id}" OR cf[10470] = "{jira_id}") '
        f"AND project in ({projects_clause}) "
        f"AND updated >= -7d ORDER BY updated DESC"
    )
    data = jira_request(jql, fields)
    issues = data.get("issues", [])
    print(f"    Found {len(issues)} tickets")

    for issue in issues:
        key = issue["key"]
        if key in jira_tickets:
            continue
        f_data = issue.get("fields", {})
        sprint_name = extract_sprint_name(f_data.get("customfield_10020"))
        if sprint_name and sprint_pattern.match(sprint_name) and not active_sprint_name:
            active_sprint_name = sprint_name

        assignee = f_data.get("assignee") or {}
        qa_contact = f_data.get("customfield_10470") or {}

        jira_tickets[key] = {
            "key": key,
            "summary": f_data.get("summary", ""),
            "status": (f_data.get("status") or {}).get("name", ""),
            "resolution": (f_data.get("resolution") or {}).get("name", ""),
            "resolutiondate": f_data.get("resolutiondate", "") or "",
            "statuscategorychangedate": f_data.get("statuscategorychangedate", "") or "",
            "issuetype": (f_data.get("issuetype") or {}).get("name", ""),
            "priority": (f_data.get("priority") or {}).get("name", ""),
            "assignee_id": assignee.get("accountId", "") if isinstance(assignee, dict) else "",
            "assignee_name": assignee.get("displayName", "") if isinstance(assignee, dict) else "",
            "qa_contact_id": qa_contact.get("accountId", "") if isinstance(qa_contact, dict) else "",
            "qa_contact_name": qa_contact.get("displayName", "") if isinstance(qa_contact, dict) else "",
            "sprint_name": sprint_name,
            "issuelinks": f_data.get("issuelinks") or [],
        }

print(f"  Total unique Jira tickets: {len(jira_tickets)}")
print(f"  Active sprint: {active_sprint_name or 'not found'}")

# Fetch sprint backlog if we found an active sprint
if active_sprint_name:
    print(f"\n=== Fetching Sprint Backlog: {active_sprint_name} ===")
    jql = f'sprint = "{active_sprint_name}" ORDER BY key ASC'
    data = jira_request(jql, fields)
    sprint_issues = data.get("issues", [])
    print(f"  Sprint backlog: {len(sprint_issues)} tickets")

    added = 0
    for issue in sprint_issues:
        key = issue["key"]
        if key in jira_tickets:
            continue
        f_data = issue.get("fields", {})
        sprint_name = extract_sprint_name(f_data.get("customfield_10020"))
        assignee = f_data.get("assignee") or {}
        qa_contact = f_data.get("customfield_10470") or {}

        jira_tickets[key] = {
            "key": key,
            "summary": f_data.get("summary", ""),
            "status": (f_data.get("status") or {}).get("name", ""),
            "resolution": (f_data.get("resolution") or {}).get("name", ""),
            "resolutiondate": f_data.get("resolutiondate", "") or "",
            "statuscategorychangedate": f_data.get("statuscategorychangedate", "") or "",
            "issuetype": (f_data.get("issuetype") or {}).get("name", ""),
            "priority": (f_data.get("priority") or {}).get("name", ""),
            "assignee_id": assignee.get("accountId", "") if isinstance(assignee, dict) else "",
            "assignee_name": assignee.get("displayName", "") if isinstance(assignee, dict) else "",
            "qa_contact_id": qa_contact.get("accountId", "") if isinstance(qa_contact, dict) else "",
            "qa_contact_name": qa_contact.get("displayName", "") if isinstance(qa_contact, dict) else "",
            "sprint_name": sprint_name or active_sprint_name,
            "issuelinks": f_data.get("issuelinks") or [],
        }
        added += 1
    print(f"  Added {added} new tickets from sprint backlog")

print(f"  Final Jira ticket count: {len(jira_tickets)}")

# Save Jira CSV
with open(CACHE_DIR / "jira-tickets.csv", "w") as f:
    f.write("key,summary,status,resolution,resolutiondate,statuscategorychangedate,issuetype,priority,assignee_id,assignee_name,qa_contact_id,qa_contact_name,sprint_name\n")
    for ticket in jira_tickets.values():
        f.write(
            f"{ticket['key']},{csv_escape(ticket['summary'])},{ticket['status']},"
            f"{ticket['resolution']},{ticket['resolutiondate']},"
            f"{ticket['statuscategorychangedate']},{ticket['issuetype']},"
            f"{ticket['priority']},{ticket['assignee_id']},"
            f"{csv_escape(ticket['assignee_name'])},{ticket['qa_contact_id']},"
            f"{csv_escape(ticket['qa_contact_name'])},{ticket['sprint_name']}\n"
        )

# --- CUSTOMER ACCOUNTS (from issue links) ---
print("\n=== Extracting Customer Accounts from Issue Links ===")
customer_cases = []
jira_base = jira_config["cloud_id"]

for key, ticket in jira_tickets.items():
    for link in ticket.get("issuelinks", []):
        link_type = link.get("type", {})
        if link_type.get("name") != "Account":
            continue
        outward = link.get("outwardIssue")
        if not outward:
            continue
        account_key = outward.get("key", "")
        if not account_key.startswith("CIPOE-"):
            continue
        customer_name = (outward.get("fields") or {}).get("summary", "")
        customer_cases.append({
            "ticket_key": key,
            "case_id": account_key,
            "case_url": f"https://{jira_base}/browse/{account_key}",
            "customer_name": customer_name,
        })

print(f"  Found {len(customer_cases)} customer accounts across {len(set(c['ticket_key'] for c in customer_cases))} tickets")

with open(CACHE_DIR / "customer-cases.csv", "w") as f:
    f.write("ticket_key,case_id,case_url,customer_name\n")
    for cc in customer_cases:
        f.write(
            f"{cc['ticket_key']},{cc['case_id']},{cc['case_url']},"
            f"{csv_escape(cc['customer_name'])}\n"
        )

# --- GITLAB ---
print("\n=== Fetching GitLab MRs ===")
gitlab_mrs = []
GITLAB_API = config.get("gitlab", {}).get("api_url", os.environ.get("GITLAB_API_URL", ""))
GITLAB_BASE = GITLAB_API.rsplit("/api/", 1)[0] + "/" if GITLAB_API else ""

for eng in engineers:
    name = eng["name"]
    username = eng["gitlab"]
    print(f"  {name} ({username})...")

    # Merged MRs
    url = (
        f"{GITLAB_API}/merge_requests?"
        f"author_username={username}&scope=all&state=merged"
        f"&updated_after={SEVEN_DAYS_AGO_ISO}&per_page=100"
    )
    data = gitlab_request(url)
    if isinstance(data, list):
        for mr in data:
            web_url = mr.get("web_url", "")
            project_path = web_url.replace(GITLAB_BASE, "").split("/-/")[0] if "/-/" in web_url else ""
            gitlab_mrs.append({
                "engineer": name,
                "iid": mr.get("iid", ""),
                "title": mr.get("title", ""),
                "project_path": project_path,
                "state": "merged",
                "created_at": mr.get("created_at", ""),
                "merged_at": mr.get("merged_at", ""),
                "web_url": web_url,
            })

    # Open MRs
    url = (
        f"{GITLAB_API}/merge_requests?"
        f"author_username={username}&scope=all&state=opened&per_page=100"
    )
    data = gitlab_request(url)
    if isinstance(data, list):
        for mr in data:
            web_url = mr.get("web_url", "")
            project_path = web_url.replace(GITLAB_BASE, "").split("/-/")[0] if "/-/" in web_url else ""
            gitlab_mrs.append({
                "engineer": name,
                "iid": mr.get("iid", ""),
                "title": mr.get("title", ""),
                "project_path": project_path,
                "state": "open",
                "created_at": mr.get("created_at", ""),
                "merged_at": "",
                "web_url": web_url,
            })

print(f"  Total GitLab MRs: {len(gitlab_mrs)}")

# Save GitLab CSV
with open(CACHE_DIR / "gitlab-mrs.csv", "w") as f:
    f.write("engineer,iid,title,project_path,state,created_at,merged_at,web_url\n")
    for mr in gitlab_mrs:
        f.write(
            f"{csv_escape(mr['engineer'])},{mr['iid']},{csv_escape(mr['title'])},"
            f"{mr['project_path']},{mr['state']},{mr['created_at']},"
            f"{mr['merged_at']},{mr['web_url']}\n"
        )

# Save last-updated timestamp
with open(CACHE_DIR / "last-updated.txt", "w") as f:
    f.write(datetime.now().isoformat() + "\n")

# Summary
print(f"\n=== Summary ===")
print(f"GitHub PRs: {len(github_prs)}")
print(f"Jira Tickets: {len(jira_tickets)}")
print(f"GitLab MRs: {len(gitlab_mrs)}")
print(f"Active Sprint: {active_sprint_name or 'not found'}")
print(f"Cache saved to: {CACHE_DIR}")

File: agents/sprint-review/scripts/fetch-data.ts
#!/usr/bin/env npx tsx
/**
 * Fetch sprint review data from Jira REST API.
 * Discovers the active sprint, fetches all issues and changelogs for cycle time.
 */

import { readFileSync, writeFileSync, mkdirSync, rmSync } from "node:fs";
import { resolve } from "node:path";

const SCRIPT_DIR = import.meta.dirname;
const DATA_DIR = resolve(SCRIPT_DIR, "..", "data");
const CACHE_DIR = resolve(DATA_DIR, "cache");
const CONFIG_PATH = resolve(DATA_DIR, "sprint-config.json");

// --- Types ---

interface Engineer {
  name: string;
  jira_account_id: string;
}

interface SprintConfig {
  board_id: number;
  sprint_name_prefix: string;
  jira: {
    cloud_id: string;
    sprint_field: string;
    story_point_field: string;
  };
  statuses: {
    in_progress: string[];
    done: string[];
  };
  engineers: Engineer[];
}

interface SprintObject {
  name?: string;
  state?: string;
  boardId?: number;
  startDate?: string;
  endDate?: string;
}

interface JiraIssue {
  key: string;
  fields?: Record<string, unknown>;
}

interface JiraSearchResult {
  issues?: JiraIssue[];
  total?: number;
  nextPageToken?: string;
}

interface ChangelogItem {
  field?: string;
  toString?: string;
}

interface ChangelogHistory {
  created?: string;
  items?: ChangelogItem[];
}

interface ChangelogResult {
  changelog?: {
    histories?: ChangelogHistory[];
  };
}

// --- Load config ---

const config = JSON.parse(readFileSync(CONFIG_PATH, "utf-8")) as SprintConfig;

// --- Auth ---

const JIRA_API_TOKEN = process.env.JIRA_API_TOKEN ?? "";
const JIRA_EMAIL = process.env.JIRA_EMAIL ?? "";

// --- Helpers ---

function str(value: unknown): string {
  if (value == null) return "";
  return `${value as string | number}`;
}

function csvEscape(value: string): string {
  value = value.replace(/"/g, '""');
  if (value.includes(",") || value.includes('"')) {
    return `"${value}"`;
  }
  return value;
}

function getAuth(): string {
  return Buffer.from(`${JIRA_EMAIL}:${JIRA_API_TOKEN}`).toString("base64");
}

async function jiraSearch(
  jql: string,
  fields: string[],
  maxResults = 100,
  nextPageToken?: string,
): Promise<JiraSearchResult> {
  const base = `https://${config.jira.cloud_id}/rest/api/3/search/jql`;
  const params = new URLSearchParams({
    jql,
    maxResults: String(maxResults),
    fields: fields.join(","),
  });
  if (nextPageToken) params.set("nextPageToken", nextPageToken);
  const url = `${base}?${params}`;
  try {
    const res = await fetch(url, {
      headers: {
        Authorization: `Basic ${getAuth()}`,
        Accept: "application/json",
      },
    });
    if (!res.ok) {
      const body = (await res.text()).slice(0, 200);
      console.error(`  Jira search error ${res.status}: ${body}`);
      return { issues: [] };
    }
    return (await res.json()) as JiraSearchResult;
  } catch (e) {
    console.error(`  Jira search error: ${String(e)}`);
    return { issues: [] };
  }
}

async function jiraGetIssue(
  key: string,
  expand: string,
  fields: string[],
): Promise<Record<string, unknown>> {
  const base = `https://${config.jira.cloud_id}/rest/api/3/issue/${key}`;
  const params = new URLSearchParams({
    expand,
    fields: fields.join(","),
  });
  const url = `${base}?${params}`;
  try {
    const res = await fetch(url, {
      headers: {
        Authorization: `Basic ${getAuth()}`,
        Accept: "application/json",
      },
    });
    if (!res.ok) {
      const body = (await res.text()).slice(0, 200);
      console.error(`  Jira issue error ${res.status} for ${key}: ${body}`);
      return {};
    }
    return (await res.json()) as Record<string, unknown>;
  } catch (e) {
    console.error(`  Jira issue error for ${key}: ${String(e)}`);
    return {};
  }
}

async function fetchAllIssues(
  jql: string,
  fields: string[],
): Promise<JiraIssue[]> {
  const allIssues: JiraIssue[] = [];
  let nextPageToken: string | undefined;
  for (;;) {
    const result = await jiraSearch(jql, fields, 100, nextPageToken);
    const issues = result.issues ?? [];
    allIssues.push(...issues);
    if (issues.length < 100 || !result.nextPageToken) break;
    nextPageToken = result.nextPageToken;
  }
  return allIssues;
}

// --- Main ---

async function main(): Promise<void> {
  if (!JIRA_API_TOKEN || !JIRA_EMAIL) {
    console.error(
      "JIRA_API_TOKEN and JIRA_EMAIL environment variables are required.",
    );
    process.exit(1);
  }

  mkdirSync(CACHE_DIR, { recursive: true });

  // Clear old cache
  for (const f of [
    "sprint-issues.csv",
    "sprint-transitions.csv",
    "sprint-changelog.csv",
    "last-updated.txt",
  ]) {
    try {
      rmSync(resolve(CACHE_DIR, f), { force: true });
    } catch {
      // ignore
    }
  }

  // --- Step 1: Discover Active Sprint ---

  console.log("=== Discovering Active Sprint ===");

  let sprintName = "";
  let sprintStart = "";
  let sprintEnd = "";

  const sprintFields = ["summary", config.jira.sprint_field];

  for (const eng of config.engineers) {
    console.log(
      `  Trying ${eng.name} (${eng.jira_account_id.slice(0, 20)}...)...`,
    );
    const jql = `sprint in openSprints() AND assignee = "${eng.jira_account_id}" ORDER BY updated DESC`;
    const result = await jiraSearch(jql, sprintFields);
    const issues = result.issues ?? [];

    for (const issue of issues) {
      const sprintField = issue.fields?.[config.jira.sprint_field];
      if (!Array.isArray(sprintField)) continue;
      const sprints = sprintField as SprintObject[];
      for (const s of sprints) {
        if (
          s.state === "active" &&
          s.boardId === config.board_id &&
          s.name?.startsWith(config.sprint_name_prefix)
        ) {
          sprintName = s.name;
          sprintStart = s.startDate ?? "";
          sprintEnd = s.endDate ?? "";
          break;
        }
      }
      if (sprintName) break;
    }
    if (sprintName) break;
  }

  if (!sprintName) {
    console.error(
      `No active sprint found for board ${config.board_id}. Verify the board ID in sprint-config.json.`,
    );
    process.exit(1);
  }

  console.log(`  Active sprint: ${sprintName}`);
  console.log(`  Start: ${sprintStart}`);
  console.log(`  End: ${sprintEnd}`);

  // --- Step 2: Fetch All Sprint Issues ---

  console.log("\n=== Fetching Sprint Issues ===");

  const issueFields = [
    "summary",
    "status",
    "assignee",
    "resolution",
    "resolutiondate",
    "issuetype",
    "priority",
    "created",
    "updated",
    config.jira.story_point_field,
    config.jira.sprint_field,
    "labels",
    "customfield_10470",
  ];

  const jql = `sprint = "${sprintName}" ORDER BY status ASC, priority DESC`;
  const allIssues = await fetchAllIssues(jql, issueFields);
  console.log(`  Found ${allIssues.length} issues`);

  if (allIssues.length === 0) {
    console.error(`Sprint '${sprintName}' returned no issues.`);
    process.exit(1);
  }

  // --- Step 3: Fetch Changelogs for Completed Issues ---

  console.log("\n=== Fetching Transition Timestamps ===");

  const doneStatuses = new Set(
    config.statuses.done.map((s) => s.toLowerCase()),
  );
  const inProgressStatuses = new Set(
    config.statuses.in_progress.map((s) => s.toLowerCase()),
  );

  const completedKeys: string[] = [];
  for (const issue of allIssues) {
    const f = issue.fields ?? {};
    const resolution = str(
      (f.resolution as Record<string, unknown> | null)?.name,
    );
    const status = str((f.status as Record<string, unknown> | null)?.name);
    if (
      resolution.toLowerCase() === "done" ||
      doneStatuses.has(status.toLowerCase())
    ) {
      completedKeys.push(issue.key);
    }
  }

  console.log(`  ${completedKeys.length} completed issues to fetch changelogs`);

  const transitions: Array<{ key: string; first_in_progress_date: string }> =
    [];

  for (const key of completedKeys) {
    const data = (await jiraGetIssue(key, "changelog", [
      "summary",
    ])) as ChangelogResult;
    const histories = data.changelog?.histories ?? [];

    let earliestDate = "";
    for (const history of histories) {
      const items = history.items ?? [];
      for (const item of items) {
        if (
          item.field === "status" &&
          item.toString &&
          inProgressStatuses.has(item.toString.toLowerCase())
        ) {
          const created = history.created ?? "";
          if (!earliestDate || created < earliestDate) {
            earliestDate = created;
          }
        }
      }
    }

    if (earliestDate) {
      transitions.push({ key, first_in_progress_date: earliestDate });
    }
  }

  console.log(
    `  Found ${transitions.length} issues with in-progress transitions`,
  );

  // --- Step 4: Save CSVs ---

  console.log("\n=== Saving CSV Files ===");

  // sprint-issues.csv
  const issueLines = [
    "key,summary,status,resolution,resolutiondate,issuetype,priority,assignee_id,assignee_name,story_points,created,updated,sprint_name,sprint_start,sprint_end,labels,qa_contact_id,qa_contact_name",
  ];
  for (const issue of allIssues) {
    const f = issue.fields ?? {};
    const assignee = (f.assignee ?? {}) as Record<string, unknown>;
    const qaContact = (f.customfield_10470 ?? {}) as Record<string, unknown>;
    const sp = f[config.jira.story_point_field];
    const labels = Array.isArray(f.labels)
      ? (f.labels as string[]).join(";")
      : "";

    issueLines.push(
      `${issue.key},${csvEscape(str(f.summary))},${str((f.status as Record<string, unknown> | null)?.name)},` +
        `${str((f.resolution as Record<string, unknown> | null)?.name)},${str(f.resolutiondate)},` +
        `${str((f.issuetype as Record<string, unknown> | null)?.name)},` +
        `${str((f.priority as Record<string, unknown> | null)?.name)},` +
        `${str(assignee?.accountId)},${csvEscape(str(assignee?.displayName))},` +
        `${sp != null ? str(sp) : ""},` +
        `${str(f.created)},${str(f.updated)},` +
        `${csvEscape(sprintName)},${sprintStart},${sprintEnd},` +
        `${csvEscape(labels)},` +
        `${str(qaContact?.accountId)},${csvEscape(str(qaContact?.displayName))}`,
    );
  }
  writeFileSync(
    resolve(CACHE_DIR, "sprint-issues.csv"),
    issueLines.join("\n") + "\n",
  );
  console.log(`  sprint-issues.csv: ${allIssues.length} issues`);

  // sprint-transitions.csv
  const transitionLines = ["key,first_in_progress_date"];
  for (const t of transitions) {
    transitionLines.push(`${t.key},${t.first_in_progress_date}`);
  }
  writeFileSync(
    resolve(CACHE_DIR, "sprint-transitions.csv"),
    transitionLines.join("\n") + "\n",
  );
  console.log(`  sprint-transitions.csv: ${transitions.length} transitions`);

  // sprint-changelog.csv — header only
  writeFileSync(
    resolve(CACHE_DIR, "sprint-changelog.csv"),
    "key,summary,status,resolution,issuetype,assignee_name,story_points,created,updated,sprint_names\n",
  );
  console.log("  sprint-changelog.csv: header only");

  // last-updated.txt
  writeFileSync(
    resolve(CACHE_DIR, "last-updated.txt"),
    new Date().toISOString() + "\n",
  );

  // --- Summary ---

  console.log("\n=== Summary ===");
  console.log(`Sprint: ${sprintName}`);
  console.log(`Issues: ${allIssues.length}`);
  console.log(`Completed: ${completedKeys.length}`);
  console.log(`Transitions: ${transitions.length}`);
  console.log(`Cache saved to: ${CACHE_DIR}`);
}

const isDirectRun = process.argv[1]?.endsWith("fetch-data.ts");
if (isDirectRun) void main();
